airflow/www_rbac/security.py

```diff
@@ -120,60 +139,270 @@
     {
         'role': 'Viewer',
         'perms': viewer_perms,
-        'vms': viewer_vms,
+        'vms': viewer_vms | dag_vms
     },
     {
         'role': 'User',
-        'perms': viewer_perms + user_perms,
-        'vms': viewer_vms + user_vms,
+        'perms': viewer_perms | user_perms | dag_perms,
+        'vms': viewer_vms | dag_vms | user_vms,
     },
     {
         'role': 'Op',
-        'perms': viewer_perms + user_perms + op_perms,
-        'vms': viewer_vms + user_vms + op_vms,
+        'perms': viewer_perms | user_perms | op_perms | dag_perms,
+        'vms': viewer_vms | dag_vms | user_vms | op_vms,
     },
 ]
 
+EXISTING_ROLES = {
+    'Admin',
+    'Viewer',
+    'User',
+    'Op',
+    'Public',
+}
+
+
+class AirflowSecurityManager(SecurityManager):
+
+    def init_role(self, role_name, role_vms, role_perms):
+        """
+        Initialize the role with the permissions and related view-menus.
+
+        :param role_name:
+        :param role_vms:
+        :param role_perms:
+        :return:
+        """
+        pvms = self.get_session.query(sqla_models.PermissionView).all()
+        pvms = [p for p in pvms if p.permission and p.view_menu]
+
+        role = self.find_role(role_name)
+        if not role:
+            role = self.add_role(role_name)
+
+        role_pvms = []
+        for pvm in pvms:
+            if pvm.view_menu.name in role_vms and pvm.permission.name in role_perms:
+                role_pvms.append(pvm)
+        role.permissions = list(set(role_pvms))
+        self.get_session.merge(role)
+        self.get_session.commit()
+
+    def get_user_roles(self, user=None):
+        """
+        Get all the roles associated with the user.
+        """
+        if user is None:
+            user = g.user
+        if user.is_anonymous():
+            public_role = appbuilder.config.get('AUTH_ROLE_PUBLIC')
+            return [appbuilder.security_manager.find_role(public_role)] \
+                if public_role else []
+        return user.roles
+
+    def get_all_permissions_views(self):
+        """
+        Returns a set of tuples with the perm name and view menu name
+        """
+        perms_views = set()
+        for role in self.get_user_roles():
+            for perm_view in role.permissions:
+                perms_views.add((perm_view.permission.name, perm_view.view_menu.name))
+        return perms_views
+
+    def get_accessible_dag_ids(self, username=None):
+        """
+        Return a set of dags that user has access to(either read or write).
+
+        :param username: Name of the user.
+        :return: A set of dag ids that the user could access.
+        """
+        if not username:
+            username = g.user
+
+        if username.is_anonymous() or 'Public' in username.roles:
+            # return an empty list if the role is public
+            return set()
+
+        roles = [role.name for role in username.roles]
+
+        if 'Admin' in roles or \
```

Comment:
NIT: You could check for set intersection here
```python
roles = {role.name for role in username.roles}
if {'Admin', 'Viewer', 'User', 'Op'} & roles:
    return dag_vms
```

Reply:
thanks